storage/src/index/storage.rs

```diff
@@ -0,0 +1,166 @@
+//! A memory efficient index structure for a journal that maps keys to locations.
+
+use crate::index::Translator;
+use commonware_runtime::Metrics;
+use prometheus_client::metrics::counter::Counter;
+use std::collections::{hash_map::Entry, HashMap};
+
+/// Each key is mapped to a `Record` that contains a potential location of the requested value in
+/// the journal, and an optional link to another potential location to allow for collisions.
+///
+/// The `Index` uses a `Translator` to transform keys into a compressed representation, resulting in
+/// non-negligible probability of collisions. Collision resolution is the responsibility of the
+/// user.
+struct Record {
+    location: u64,
+
+    next: Option<Box<Record>>,
+}
+
+pub struct LocationIterator<'a> {
+    next: Option<&'a Record>,
+}
+
+impl LocationIterator<'_> {
+    /// Create a `LocationIterator` that returns no items.
+    fn empty() -> Self {
+        LocationIterator { next: None }
+    }
+}
+
+impl Iterator for LocationIterator<'_> {
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.next {
+            Some(next) => {
+                let loc = next.location;
+                self.next = next.next.as_deref();
+                Some(loc)
+            }
+            None => None,
+        }
+    }
+}
+
+impl Record {
+    fn iter(&self) -> LocationIterator {
+        LocationIterator { next: Some(self) }
+    }
+}
+
+pub struct Index<T: Translator, E: Metrics> {
+    translator: T,
+
+    // We store the first location associated with a key within the HashMap entry itself to
+    // significantly reduce the number of random reads we need to do on the heap.
+    map: HashMap<T::Key, Record>,
+
+    collisions: Counter,
+    keys_pruned: Counter,
+
+    _metrics: std::marker::PhantomData<E>,
+}
+
+impl<T: Translator, E: Metrics> Index<T, E> {
+    /// Create a new index.
+    pub fn init(context: E, translator: T) -> Self {
+        let s = Self {
+            translator,
+            map: HashMap::new(),
+            collisions: Counter::default(),
+            keys_pruned: Counter::default(),
+            _metrics: std::marker::PhantomData,
+        };
+        context.register(
+            "keys_pruned",
+            "Number of keys pruned",
+            s.keys_pruned.clone(),
+        );
+
+        s
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn collisions(&self) -> u64 {
```

Comment:
I think we can remove this getter now that collisions is a metric?

Reply:
I think we use this in some logging above...we can drop that imo.

Reply:
yup done